File: tests/ten_runtime/integration/go/prepare_to_stop_go/prepare_to_stop_go_app/ten_packages/extension/extension_b/extension.go
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package default_extension_go

import (
	"fmt"

	ten "ten_framework/ten_runtime"
)

type bExtension struct {
	ten.DefaultExtension
	stopChan chan struct{}
}

func NewBExtension(name string) ten.Extension {
	return &bExtension{
		stopChan: make(chan struct{}),
	}
}

func (p *bExtension) OnCmd(
	tenEnv ten.TenEnv,
	cmd ten.Cmd,
) {
	go func() {
		cmdName, _ := cmd.GetName()
		tenEnv.Log(
			ten.LogLevelInfo,
			"receive command: "+
				cmdName,
		)
		if cmdName == "start" {
			tenEnv.SendCmd(cmd, func(r ten.TenEnv, cs ten.CmdResult, e error) {
				r.ReturnResult(cs, nil)
			})
		} else if cmdName == "stop" {
			tenEnv.SendCmd(cmd, func(r ten.TenEnv, cs ten.CmdResult, e error) {
				r.ReturnResult(cs, nil)

				close(p.stopChan)
				tenEnv.Log(ten.LogLevelInfo, "Stop command is processed.")
			})
		} else {
			cmdResult, _ := ten.NewCmdResult(ten.StatusCodeError, cmd)
			cmdResult.SetPropertyString("detail", "unknown cmd")
			tenEnv.ReturnResult(cmdResult, nil)
		}
	}()
}

func (p *bExtension) OnStop(tenEnv ten.TenEnv) {
	go func() {
		tenEnv.Log(ten.LogLevelDebug, "OnStop")

		// Wait until the stop command is received and processed.
		<-p.stopChan

		tenEnv.Log(
			ten.LogLevelInfo,
			"Stop command processed. Now calling OnStopDone.",
		)
		tenEnv.OnStopDone()
	}()
}

func init() {
	fmt.Println("call init")

	// Register addon
	err := ten.RegisterAddonAsExtension(
		"extension_b",
		ten.NewDefaultExtensionAddon(NewBExtension),
	)
	if err != nil {
		fmt.Println("register addon failed", err)
	}
}

File: tests/ten_runtime/integration/python/go_app_async_extension_python/go_app_async_extension_python_app/ten_packages/extension/default_extension_python/extension.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#

# import debugpy
# debugpy.listen(5678)
# debugpy.wait_for_client()

import asyncio
from ten_runtime import (
    AsyncExtension,
    AsyncTenEnv,
    Cmd,
    StatusCode,
    CmdResult,
    LogLevel,
)


class DefaultExtension(AsyncExtension):
    async def on_configure(self, ten_env: AsyncTenEnv) -> None:
        ten_env.log(LogLevel.DEBUG, "on_init")

        await ten_env.init_property_from_json('{"testKey": "testValue"}')

        await asyncio.sleep(0.5)

    async def on_start(self, ten_env: AsyncTenEnv) -> None:
        ten_env.log(LogLevel.DEBUG, "on_start")

        await asyncio.sleep(0.5)

        await ten_env.set_property_from_json("testKey2", '"testValue2"')
        testValue, _ = await ten_env.get_property_to_json("testKey")
        testValue2, _ = await ten_env.get_property_to_json("testKey2")
        ten_env.log(
            LogLevel.INFO, f"testValue: {testValue}, testValue2: {testValue2}"
        )

    async def on_stop(self, ten_env: AsyncTenEnv) -> None:
        ten_env.log(LogLevel.DEBUG, "on_stop")

        await asyncio.sleep(0.5)

    async def on_deinit(self, ten_env: AsyncTenEnv) -> None:
        ten_env.log(LogLevel.DEBUG, "on_deinit")

        await asyncio.sleep(0.5)

    async def greeting(self, ten_env: AsyncTenEnv) -> CmdResult | None:
        await asyncio.sleep(1)

        new_cmd = Cmd.create("greeting")
        result, err = await ten_env.send_cmd(new_cmd)
        if err is not None:
            ten_env.log(
                LogLevel.FATAL, f"greeting error: {err.error_message()}"
            )

        return result

    async def on_cmd(self, ten_env: AsyncTenEnv, cmd: Cmd) -> None:
        cmd_json, _ = cmd.get_property_to_json()
        ten_env.log(LogLevel.DEBUG, "on_cmd: " + cmd_json)

        new_cmd = Cmd.create("hello")
        new_cmd.set_property_from_json("test", '"testValue2"')
        test_value, _ = new_cmd.get_property_to_json("test")
        ten_env.log(LogLevel.INFO, f"on_cmd test_value: {test_value}")

        await asyncio.sleep(0.5)

        result, _ = await ten_env.send_cmd(new_cmd)
        assert result is not None

        statusCode = result.get_status_code()
        detail, _ = result.get_property_string("detail")
        ten_env.log(
            LogLevel.INFO,
            f"check_hello: status: {str(statusCode)}, detail: {detail}",
        )

        greeting_tasks = [self.greeting(ten_env) for _ in range(100)]

        results = await asyncio.gather(*greeting_tasks)

        for result in results:
            if result is None:
                ten_env.log(LogLevel.FATAL, "check_hello: result is None")
                assert False

            statusCode = result.get_status_code()
            if statusCode != StatusCode.OK:
                ten_env.log(
                    LogLevel.FATAL, f"check_hello: status: {str(statusCode)}"
                )
                assert False

        respCmd = CmdResult.create(StatusCode.OK, cmd)
        respCmd.set_property_string("detail", "received response")
        ten_env.log(LogLevel.INFO, "create respCmd")

        await ten_env.return_result(respCmd)
